### PR TITLE
feat: allow secret parameter to be a function to support dynamic secrets

### LINKED ISSUE #65: Support dynamic secrets
As a user of the package, I want to be able to supply a function as the `secret` parameter so I can dynamically determine the `secret` based off the `req` object or other parameters. This should support `async` functions and must return a `string` or `undefined` (which will then throw an error in the package)

### PR DESCRIPTION
**Pull request by connorjburton**

### CODE REVIEW

**Comment by connorjburton** [comment]
@neslisahgamze  Please review/try this branch and ensure it fits your needs. I still need to update documentation, version bump & changelog but it should be complete functionality wise.

**Comment by connorjburton** [comment]
@Jimsalad Quick code review if you don't mind and have the time?

**Review by neslisahgamze — APPROVED** [review]

**Comment by neslisahgamze** [comment]
> @neslisahgamze Please review/try this branch and ensure it fits your needs. I still need to update documentation, version bump & changelog but it should be complete functionality wise.

I checked this PR on my code and it works. Thank you 🎉

File: README.md [hunk 1]
```diff
@@ -100,13 +100,23 @@ app.use((error, req, res, next) => {
...
-
-  else {
-    // ... handle other errors
-  }
 })
 ```
 
+## Dynamic Secret
+
+From [8.2.0](https://github.com/connorjburton/hmac-auth-express/releases/tag/v8.2.0) onwards you can supply a function as your `secret` parameter. This function accepts 1 parameter, being an `express.Request` object. This function can be `async`. You can use this feature to dynamically determine the secret of the request, for example if you have different HMAC secrets depending on the URL of the request.
```
Comment (james-salloway) [inline README.md]:
```suggestion
From [8.2.0](https://github.com/connorjburton/hmac-auth-express/releases/tag/v8.2.0) onwards you can alternatively supply a function as your `secret` parameter. This function accepts 1 parameter, being an `express.Request` object. This function can be `async`. You can use this feature to dynamically determine the secret of the request, for example if you have different HMAC secrets depending on the URL of the request.
```

File: src/index.ts [hunk 2]
```diff
@@ -40,12 +42,24 @@ export function generate(secret: string, algorithm: string = defaults.algorithm,
...
     validateArguments(secret, mergedOpts);
 
-    return function(request: Request, _, next: NextFunction): void {
+    return async function(request: Request, _, next: NextFunction): Promise<void> {
+        // we have to create a scoped secret per request, if we were to reassign the original `secret` variable
+        // the next request that comes in will no longer be the secret function
+        // this means we have to explicitly check it's not undefined too
+        const scopedSecret = await determineSecret(secret, request);
+        if (typeof scopedSecret !== 'string' || scopedSecret.length === 0) {
+            return next(new AuthError(`Invalid secret. Expected non-empty string but got '${scopedSecret}' (type: ${typeof scopedSecret})`));
```
Comment (james-salloway) [inline src/index.ts:60]: If you're concerned with `arrays` and `Promises` reporting correctly from the `typeof scopedSecret` error statement you will need to handle these independently - `arrays` will return as `object` and Promises will return as `function`.

**Review by james-salloway — CHANGES_REQUESTED** [review]
Added a couple of nitpicks:

`index.ts` - return values from the `AuthError` return will not provide accurate `typeof` statements if certain data types such as `Promises` or `arrays` are passed in. This can be ignored if you're happy to stick with explicit JavaScript data structures.

`README.md` - proper nitpick, just changing some wording around what parameters you can provide.

Overall looks good though!

**Comment by connorjburton** [comment]
@neslisahgamze This is now released on NPM as `8.2.0`, thanks for your help.

**Pull request merged**